File: design-pattern-php/Factory/FactoryMethod.php
<?php

/**
 * 工厂方法模式，和简单工厂的差别是工厂方法模式是具体的工厂产生具体的产品
 */

require_once __DIR__ . "/Animal.php";

/**
 * 动物工厂接口
 */
interface AnimalFactory
{
    public static function make(): Animal;
}

class CatFactory implements AnimalFactory
{
    public static function make(): Animal
    {
        return new Cat;
    }
}

class DogFactory implements AnimalFactory
{
    public static function make(): Animal
    {
        return new Dog;
    }
}

$cat = CatFactory::make();
$dog = DogFactory::make();

var_dump($cat, $dog);

// F:\Code\code\design-pattern\Factory\FactoryMethod.php:36:
// class Cat#1 (1) {
//   public $name =>
//   string(3) "Cat"
// }
// F:\Code\code\design-pattern\Factory\FactoryMethod.php:36:
// class Dog#2 (1) {
//   public $name =>
//   string(3) "Dog"
// }

File: design-pattern-java/src/main/java/me/ixk/design_pattern/factory/factory_method/FactoryMethod.java
package me.ixk.design_pattern.factory.factory_method;

import me.ixk.design_pattern.factory.animal.Animal;

/**
 * 简单工厂：通过多态来拆分简单工厂的 if 判断
 * <p>
 * 优点：当需要增加新的实例的时候无效修改原始代码，只需要另外创建一个对应的工厂
 * <p>
 * 缺点：工厂会因需要创建的类越来越多而逐渐增多（此时可以通过创建一个工厂的简单工厂来解决这个问题）
 */
public interface FactoryMethod<T extends Animal> {
    T createAnimal();
}
